File: src/CodeBeaker.Core/Docker/DockerExecutor.cs
using Docker.DotNet;
using Docker.DotNet.Models;
using CodeBeaker.Core.Models;
using System.Diagnostics;
using System.Text;

namespace CodeBeaker.Core.Docker;

/// <summary>
/// Docker 컨테이너 실행기
/// </summary>
public sealed class DockerExecutor
{
    private readonly DockerClient _client;

    public DockerExecutor(string dockerHost = "unix:///var/run/docker.sock")
    {
        // Windows에서는 npipe 사용
        if (OperatingSystem.IsWindows())
        {
            dockerHost = "npipe://./pipe/docker_engine";
        }

        _client = new DockerClientConfiguration(new Uri(dockerHost))
            .CreateClient();
    }

    /// <summary>
    /// Docker 컨테이너에서 코드 실행
    /// </summary>
    public async Task<ExecutionResult> ExecuteAsync(
        string image,
        string[] command,
        string workspaceDir,
        ExecutionConfig config,
        CancellationToken cancellationToken = default)
    {
        var sw = Stopwatch.StartNew();
        var executionId = Guid.NewGuid().ToString();

        try
        {
            // Create container
            var container = await CreateContainerAsync(image, command, workspaceDir, config, cancellationToken);

            try
            {
                // Start container
                await _client.Containers.StartContainerAsync(container.ID, new ContainerStartParameters(), cancellationToken);

                // Wait for container with timeout
                var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                timeoutCts.CancelAfter(TimeSpan.FromSeconds(config.Timeout));

                ContainerWaitResponse? waitResponse = null;
                var timedOut = false;

                try
                {
                    waitResponse = await _client.Containers.WaitContainerAsync(container.ID, timeoutCts.Token);
                }
                catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
                {
                    // Timeout occurred
                    timedOut = true;

                    // Kill the container
                    try
                    {
                        await _client.Containers.KillContainerAsync(container.ID, new ContainerKillParameters(), cancellationToken);
                    }
                    catch
                    {
                        // Ignore kill errors
                    }
                }

                // Get logs
                var (stdout, stderr) = await GetLogsAsync(container.ID, cancellationToken);

                sw.Stop();

                if (timedOut)
                {
                    return new ExecutionResult
                    {
                        ExecutionId = executionId,
                        Status = "failed",
                        ExitCode = 124, // Timeout exit code
                        Stdout = stdout,
                        Stderr = stderr,
                        DurationMs = sw.ElapsedMilliseconds,
                        Timeout = true,
                        ErrorType = "timeout_error",
                        CompletedAt = DateTime.UtcNow
                    };
                }

                var exitCode = (int)(waitResponse?.StatusCode ?? 0);

                return new ExecutionResult
                {
                    ExecutionId = executionId,
                    Status = exitCode == 0 ? "completed" : "failed",
                    ExitCode = exitCode,
                    Stdout = stdout,
                    Stderr = stderr,
                    DurationMs = sw.ElapsedMilliseconds,
                    Timeout = false,
                    ErrorType = exitCode != 0 ? DetermineErrorType(stderr) : null,
                    CompletedAt = DateTime.UtcNow
                };
            }
            finally
            {
                // Cleanup container
                try
                {
                    await _client.Containers.RemoveContainerAsync(
                        container.ID,
                        new ContainerRemoveParameters { Force = true },
                        cancellationToken);
                }
                catch
                {
                    // Ignore cleanup errors
                }
            }
        }
        catch (Exception ex)
        {
            sw.Stop();
            return new ExecutionResult
            {
                ExecutionId = executionId,
                Status = "failed",
                ExitCode = -1,
                Stdout = string.Empty,
                Stderr = $"Docker execution error: {ex.Message}",
                DurationMs = sw.ElapsedMilliseconds,
                Timeout = false,
                ErrorType = "docker_error",
                CompletedAt = DateTime.UtcNow
            };
        }
    }

    private async Task<CreateContainerResponse> CreateContainerAsync(
        string image,
        string[] command,
        string workspaceDir,
        ExecutionConfig config,
        CancellationToken cancellationToken)
    {
        var hostConfig = new HostConfig
        {
            // Resource limits
            Memory = config.MemoryLimit * 1024L * 1024L, // Convert MB to bytes
            NanoCPUs = (long)(config.CpuLimit * 1_000_000_000L),

            // Security
            ReadonlyRootfs = config.ReadOnlyFilesystem,
            NetworkMode = config.DisableNetwork ? "none" : "bridge",

            // Workspace mount
            Binds = new List<string>
            {
                $"{Path.GetFullPath(workspaceDir)}:/workspace:rw"
            },

            // Tmpfs for writable temp directory
            Tmpfs = new Dictionary<string, string>
            {
                { "/tmp", "rw,noexec,nosuid,size=100m" }
            }
        };

        var containerConfig = new CreateContainerParameters
        {
            Image = image,
            Cmd = command,
            WorkingDir = "/workspace",
            HostConfig = hostConfig,
            Env = config.Environment.Select(kv => $"{kv.Key}={kv.Value}").ToList(),
            AttachStdout = true,
            AttachStderr = true,
            Tty = false
        };

        return await _client.Containers.CreateContainerAsync(containerConfig, cancellationToken);
    }

    private async Task<(string stdout, string stderr)> GetLogsAsync(
        string containerId,
        CancellationToken cancellationToken)
    {
        var parameters = new ContainerLogsParameters
        {
            ShowStdout = true,
            ShowStderr = true,
            Follow = false
        };

        var stream = await _client.Containers.GetContainerLogsAsync(
            containerId,
            false, // tty=false for proper demultiplexing
            parameters,
            cancellationToken);

        var stdout = new StringBuilder();
        var stderr = new StringBuilder();

        // MultiplexedStream provides automatic demultiplexing
        using var stdoutStream = new MemoryStream();
        using var stderrStream = new MemoryStream();
        using var stdinStream = Stream.Null; // No stdin needed

        await stream.CopyOutputToAsync(stdinStream, stdoutStream, stderrStream, cancellationToken);

        stdoutStream.Position = 0;
        stderrStream.Position = 0;

        using var stdoutReader = new StreamReader(stdoutStream);
        using var stderrReader = new StreamReader(stderrStream);

        var stdoutText = await stdoutReader.ReadToEndAsync();
        var stderrText = await stderrReader.ReadToEndAsync();

        return (stdoutText, stderrText);
    }

    private static string? DetermineErrorType(string stderr)
    {
        if (string.IsNullOrWhiteSpace(stderr))
        {
            return "runtime_error";
        }

        var lowerStderr = stderr.ToLowerInvariant();

        if (lowerStderr.Contains("syntaxerror") || lowerStderr.Contains("syntax error"))
        {
            return "syntax_error";
        }

        if (lowerStderr.Contains("timeout") || lowerStderr.Contains("timed out"))
        {
            return "timeout_error";
        }

        if (lowerStderr.Contains("memory") || lowerStderr.Contains("out of memory"))
        {
            return "memory_error";
        }

        return "runtime_error";
    }

    public void Dispose()
    {
        _client?.Dispose();
    }
}

File: src/runtime/docker_executor.py
"""
Phase 1.4: Docker 컨테이너 기반 실행기

Docker SDK를 사용하여 격리된 컨테이너에서 코드 실행
"""

import os
import tempfile
import time
from pathlib import Path
from typing import Optional

import docker
from docker.errors import ContainerError, ImageNotFound, APIError

from src.common.models import ExecutionConfig, ExecutionResult


class DockerExecutor:
    """Docker 컨테이너 기반 코드 실행기"""

    def __init__(self, image: str = "codebeaker-python:latest"):
        """
        Args:
            image: 사용할 Docker 이미지 이름
        """
        self.image = image
        try:
            self.client = docker.from_env()
            # 이미지 존재 확인
            self.client.images.get(self.image)
        except ImageNotFound:
            raise RuntimeError(
                f"Docker image '{self.image}' not found. "
                f"Please build it first with: "
                f"docker build -t {self.image} docker/runtimes/python/"
            )
        except APIError as e:
            raise RuntimeError(f"Docker API error: {e}")

    def execute_python(
        self,
        code: str,
        config: ExecutionConfig = ExecutionConfig(),
    ) -> ExecutionResult:
        """
        Python 코드를 Docker 컨테이너에서 실행

        Args:
            code: 실행할 Python 코드
            config: 실행 설정

        Returns:
            ExecutionResult: 실행 결과
        """
        # 임시 디렉토리에 코드 작성
        with tempfile.TemporaryDirectory() as tmpdir:
            code_path = Path(tmpdir) / "code.py"
            code_path.write_text(code, encoding="utf-8")

            start_time = time.time()

            try:
                # Docker 컨테이너 생성 및 시작
                container = self.client.containers.create(
                    image=self.image,
                    command=["python", "/workspace/code.py"],
                    volumes={tmpdir: {"bind": "/workspace", "mode": "ro"}},
                    # 리소스 제한
                    mem_limit=config.memory_limit,
                    nano_cpus=int(config.cpu_limit * 1e9),
                    # 보안 옵션
                    network_disabled=not config.network_enabled,
                    # 실행 옵션
                    detach=True,
                )

                try:
                    # 컨테이너 시작
                    container.start()

                    # 타임아웃과 함께 대기
                    exit_status = container.wait(timeout=config.timeout)

                    # 로그 가져오기
                    stdout = container.logs(stdout=True, stderr=False)
                    stderr = container.logs(stdout=False, stderr=True)

                    duration_ms = int((time.time() - start_time) * 1000)

                    # exit_status는 딕셔너리 형태: {'StatusCode': 0}
                    exit_code = exit_status.get("StatusCode", -1)

                    return ExecutionResult(
                        stdout=stdout.decode("utf-8") if stdout else "",
                        stderr=stderr.decode("utf-8") if stderr else "",
                        exit_code=exit_code,
                        duration_ms=duration_ms,
                        timeout=False,
                    )

                except Exception as wait_error:
                    # 타임아웃 또는 실행 에러
                    duration_ms = int((time.time() - start_time) * 1000)

                    # 컨테이너 강제 종료
                    try:
                        container.kill()
                    except Exception:
                        pass

                    # 타임아웃 체크
                    if "timed out" in str(wait_error).lower() or "timeout" in str(wait_error).lower():
                        return ExecutionResult(
                            stdout="",
                            stderr=f"Execution timeout after {config.timeout} seconds",
                            exit_code=-1,
                            duration_ms=duration_ms,
                            timeout=True,
                            error_type="TimeoutError",
                        )

                    # 기타 에러
                    return ExecutionResult(
                        stdout="",
                        stderr=str(wait_error),
                        exit_code=-1,
                        duration_ms=duration_ms,
                        error_type=type(wait_error).__name__,
                    )

                finally:
                    # 컨테이너 제거
                    try:
                        container.remove(force=True)
                    except Exception:
                        pass

            except Exception as e:
                # 컨테이너 생성 실패
                duration_ms = int((time.time() - start_time) * 1000)

                return ExecutionResult(
                    stdout="",
                    stderr=f"Failed to create container: {str(e)}",
                    exit_code=-1,
                    duration_ms=duration_ms,
                    error_type=type(e).__name__,
                )

    def cleanup(self) -> None:
        """Docker 클라이언트 정리"""
        if hasattr(self, "client"):
            self.client.close()

    def __enter__(self):
        """컨텍스트 매니저 진입"""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """컨텍스트 매니저 종료"""
        self.cleanup()
